py/objstr.c

```diff
@@ -1645,6 +1645,11 @@ STATIC mp_obj_t str_modulo_format(mp_obj_t pattern, size_t n_args, const mp_obj_
         }
     }
 
+    if (dict != MP_OBJ_NULL) {
+        // if `dict` exists, then it was the only element in `args` and it was consumed; either positionally, or
+        // as a map of named args, even if none were actually referenced by `pattern`.
+        arg_i = 1;
```

Comment:
If dict is supplied then n_args=1, so it would be simpler to write the below if condition as:
```c
if (dict == MP_OBJ_NULL && arg_i != n_args) {
```

Reply:
> If dict is supplied then n_args=1, so it would be simpler to write the below if condition as:

Right. It is logically the same but I thought the reasoning is better shown by setting `arg_i` to indicate that the dict was consumed, same as through the path where there are references to dict keys in the fmt string.

Note that the (convoluted, IMHO) behavior of CPython *absolutely requires* that `arg_i` be set to 1 (= dict consumed positionally) when performing a successful lookup in the dict... It affects behavior if there are subsequent positional arg references.

Because this check for `arg_i == nargs` is the last thing the function does, there is really no necessity to update `arg_i` "correctly". 

Anyway since efficiency is more important in embedded code and this saves an asm instruction, you're right. I'll make the change.